### ISSUE TITLE
IPP-DEV-09-002: Create a simple weather data predictor class that uses the KNN algorithm

### ISSUE THREAD

**Issue by labbenchstudios** [programming-in-python/ipp-exercise-tasks#85]
# Description
- Using instructions from the in-class lecture, create a module and class that uses the recently implemented KNN algorithm to predict weather changes.
- This module will need to implement the functionality listed below under Actions.

### Review the README
- Please see [README.md](https://github.com/programming-in-python/ipp-exercise-tasks) for further information on, and use of, this content.
- License for embedded documentation and source codes: [IPP-DOC-LIC](https://github.com/programming-in-python/ipp-exercise-tasks/issues/2)

### Estimated effort may vary greatly
- The estimated level of effort for this exercise shown in the 'Estimate' section below is a very rough approximation. The actual level of effort may vary greatly depending on your development and test environment, experience with the requisite technologies, and many other factors.

### Reminders on System Configuration:
- Make Sure Your System is Setup for Python and This Course
  - See [IPP-CFG-01-001](https://github.com/programming-in-python/ipp-exercise-tasks/issues/29)

- Make Sure PYTHONPATH is Set Correctly
  - Whether running Python tests within your IDE or from the command line, you must set the PYTHONPATH environment variable in every execution environment (e.g., every terminal you launch) when attempting to run any of your scripts and their tests or the IPP test app from the command line. The IPP source and test paths will be as follows:
    - {your IPP source code path}
    - {your IPP source code path}/tests
  - As a reminder, you can include the PYTHONPATH environment variable for your environment within the activation script for your virtual environment, which will set PYTHONPATH automatically whenever your virtual environment is launched

- See [IPP-DEV-01-001](https://github.com/programming-in-python/ipp-exercise-tasks/issues/40) for details.

### Step 1: In the `IPP_HOME/exercises/labmodule09` path, create a new module named `SimpleWeatherPredictor.py` and add the following:
  - Declare the import statements
  - Declare the class to be named `SimpleWeatherPredictor`.
- Here's a BRIEF example of the class definition with its important statements:

```python
from typing import List, Dict, Optional

from ipp.exercises.labmodule09.SimpleKnnAlgorithm import SimpleKnnAlgorithm

class SimpleWeatherPredictor:
```

### Step 2: Create the constructor
- Create the constructor
  - This will perform a few important tasks
    - Sets the number of neighbors to consider ('k') - this will get passed to `SimpleKnnAlgorithm`
    - Initializes a new Knn algorithm instance for each weather value to predict
  - Here's an implementation approach for the constructor:

```python
    def __init__(self, k: int = 3):
        self.k = k
        
        # Create separate KNN models for each weather metric
        self.tempPredictor = SimpleKnnAlgorithm(k)
        self.humidityPredictor = SimpleKnnAlgorithm(k)
        self.pressurePredictor = SimpleKnnAlgorithm(k)
        
        self.isTrained = False
        self.trainingCount = 0
```

### Step 3: Create a method to prepare training data from captured weather history using a sliding window technique
- This method looks a bit complicated, but is really just looking at the two previous weather property readings to help predict the next value
- Here's one implementation approach:

```python
    def prepareTrainingData(self, weatherHistory: List[Dict]) -> Dict[str, List]:
        if len(weatherHistory) < 3:
            print("Need at least 3 records for training")
            return {}
        
        tempData = []
        humidityData = []
        pressureData = []
        
        # Create training samples using sliding window
        # Use previous 2 readings to predict next value
        for i in range(2, len(weatherHistory)):
            prev2 = weatherHistory[i - 2]
            prev1 = weatherHistory[i - 1]
            current = weatherHistory[i]
            
            # Temperature training data
            if all(r.get('temperature') is not None for r in [prev2, prev1, current]):
                features = [
                    prev2['temperature'],
                    prev1['temperature']
                ]
                target = current['temperature']
                tempData.append((features, target))
            
            # Humidity training data
            if all(r.get('humidity') is not None for r in [prev2, prev1, current]):
                features = [
                    prev2['humidity'],
                    prev1['humidity']
                ]
                target = current['humidity']
                humidityData.append((features, target))
            
            # Pressure training data
            if all(r.get('pressure') is not None for r in [prev2, prev1, current]):
                features = [
                    prev2['pressure'],
                    prev1['pressure']
                ]
                target = current['pressure']
                pressureData.append((features, target))
        
        return {
            'temperature': tempData,
            'humidity': humidityData,
            'pressure': pressureData
        }
```

### Step 4: Create the training method that will train all predictors for the various weather data properties
- Here's an example implementation:

```python
    def train(self, weatherHistory: List[Dict]) -> bool:
        # Prepare training data
        trainingData = self.prepareTrainingData(weatherHistory)
        
        if not trainingData:
            return False
        
        # Train each predictor
        trainedModels = 0
        
        if 'temperature' in trainingData and trainingData['temperature']:
            self.tempPredictor.setTrainingData(trainingData['temperature'])
            trainedModels += 1
            print(f"Temperature predictor trained with {len(trainingData['temperature'])} samples")
        
        if 'humidity' in trainingData and trainingData['humidity']:
            self.humidityPredictor.setTrainingData(trainingData['humidity'])
            trainedModels += 1
            print(f"Humidity predictor trained with {len(trainingData['humidity'])} samples")
        
        if 'pressure' in trainingData and trainingData['pressure']:
            self.pressurePredictor.setTrainingData(trainingData['pressure'])
            trainedModels += 1
            print(f"Pressure predictor trained with {len(trainingData['pressure'])} samples")
        
        # Mark as trained if at least one model is trained
        if trainedModels > 0:
            self.isTrained = True
            self.trainingCount += 1
            print(f"Weather predictor training complete (iteration {self.trainingCount})")
            return True
        
        return False
```

### Step 5: Create a method to predict the next weather values for each weather property using recent data
- Here's an example implementation:

```python
    def predict(self, recentData: List[Dict]) -> Dict[str, Optional[float]]:
        predictions = {
            'temperature': None,
            'humidity': None,
            'pressure': None
        }
        
        if not self.isTrained:
            print("Predictors not trained yet")
            return predictions
        
        if len(recentData) < 2:
            print("Need at least 2 recent records for prediction")
            return predictions
        
        # Get last 2 records
        prevRecord2 = recentData[-2]
        prevRecord1 = recentData[-1]
        
        # Predict temperature
        if prevRecord2.get('temperature') is not None and prevRecord1.get('temperature') is not None:
            features = [prevRecord2['temperature'], prevRecord1['temperature']]
            predictions['temperature'] = self.tempPredictor.predictRegression(features)
        
        # Predict humidity
        if prevRecord2.get('humidity') is not None and prevRecord1.get('humidity') is not None:
            features = [prevRecord2['humidity'], prevRecord1['humidity']]
            predictions['humidity'] = self.humidityPredictor.predictRegression(features)
        
        # Predict pressure
        if prevRecord2.get('pressure') is not None and prevRecord1.get('pressure') is not None:
            features = [prevRecord2['pressure'], prevRecord1['pressure']]
            predictions['pressure'] = self.pressurePredictor.predictRegression(features)
        
        return predictions
```

### Step 6: Finally, create a method that will evaluate the prediction accuracy for each weather property.
- Here's an implementation example:
```python
    def evaluatePredictions(self, predictions: Dict, actual: Dict) -> Dict[str, Dict]:
        evaluation = {}
        
        # Evaluate temperature
        if predictions.get('temperature') and actual.get('temperature'):
            evaluation['temperature'] = self.tempPredictor.evaluatePrediction(
                actual['temperature'],
                predictions['temperature']
            )
        
        # Evaluate humidity
        if predictions.get('humidity') and actual.get('humidity'):
            evaluation['humidity'] = self.humidityPredictor.evaluatePrediction(
                actual['humidity'],
                predictions['humidity']
            )
        
        # Evaluate pressure
        if predictions.get('pressure') and actual.get('pressure'):
            evaluation['pressure'] = self.pressurePredictor.evaluatePrediction(
                actual['pressure'],
                predictions['pressure']
            )
        
        return evaluation
```

## Estimate
- Medium

### Create the test module `test_SimpleWeatherPredictor.py` within the `tests/labmodule09` path, and the class `SimpleWeatherPredictorTest(unittest.TestCase)`.
- Here is the code for the test. You may add additional tests as well:

```python
import unittest
from ipp.exercises.labmodule09.SimpleWeatherPredictor import SimpleWeatherPredictor

class SimpleWeatherPredictorTest(unittest.TestCase):
    """
    Unit tests for SimpleWeatherPredictor class.
    Tests basic functionality of weather prediction using KNN algorithm.
    """
    
    def setUp(self):
        """
        Set up test fixtures before each test method.
        Creates sample weather history data for testing.
        """
        self.predictor = SimpleWeatherPredictor(k = 3)
        
        # Create sample weather history with at least 3 records
        self.weatherHistory = [
            {'temperature': 20.0, 'humidity': 65.0, 'pressure': 1013.0},
            {'temperature': 22.0, 'humidity': 60.0, 'pressure': 1015.0},
            {'temperature': 24.0, 'humidity': 55.0, 'pressure': 1017.0},
            {'temperature': 26.0, 'humidity': 50.0, 'pressure': 1019.0},
            {'temperature': 28.0, 'humidity': 45.0, 'pressure': 1021.0}
        ]
    
    def testPrepareTrainingData(self):
        """
        Test that training data is prepared correctly from weather history.
        Should create feature-target pairs using sliding window approach.
        """
        trainingData = self.predictor.prepareTrainingData(self.weatherHistory)
        
        # Should have data for all three metrics
        self.assertIn('temperature', trainingData)
        self.assertIn('humidity', trainingData)
        self.assertIn('pressure', trainingData)
        
        # With 5 records, we should have 3 training samples (records 3-5)
        self.assertEqual(len(trainingData['temperature']), 3)
        self.assertEqual(len(trainingData['humidity']), 3)
        self.assertEqual(len(trainingData['pressure']), 3)
    
    def testTrainModel(self):
        """
        Test that the predictor can be trained successfully.
        After training, isTrained flag should be True.
        """
        success = self.predictor.train(self.weatherHistory)
        
        self.assertTrue(success)
        self.assertTrue(self.predictor.isTrained)
        self.assertEqual(self.predictor.trainingCount, 1)
    
    def testPredictAfterTraining(self):
        """
        Test that predictions can be made after training.
        Predictions should return numeric values for all metrics.
        """
        self.predictor.train(self.weatherHistory)
        
        # Use last 2 records for prediction
        recentData = self.weatherHistory[-2:]
        predictions = self.predictor.predict(recentData)
        
        # Should have predictions for all metrics
        self.assertIsNotNone(predictions['temperature'])
        self.assertIsNotNone(predictions['humidity'])
        self.assertIsNotNone(predictions['pressure'])
        
        # Predictions should be numeric
        self.assertIsInstance(predictions['temperature'], (int, float))
        self.assertIsInstance(predictions['humidity'], (int, float))
        self.assertIsInstance(predictions['pressure'], (int, float))
    
    def testPredictWithoutTraining(self):
        """
        Test that prediction fails gracefully when model is not trained.
        Should return None values for all predictions.
        """
        recentData = self.weatherHistory[-2:]
        predictions = self.predictor.predict(recentData)
        
        # Should return None for all predictions when not trained
        self.assertIsNone(predictions['temperature'])
        self.assertIsNone(predictions['humidity'])
        self.assertIsNone(predictions['pressure'])
    
    def testEvaluatePredictions(self):
        """
        Test that prediction evaluation produces error metrics.
        Should return evaluation dictionary with error measurements.
        """
        self.predictor.train(self.weatherHistory)
        
        # Make predictions
        recentData = self.weatherHistory[-2:]
        predictions = self.predictor.predict(recentData)
        
        # Actual values (last record in history)
        actual = self.weatherHistory[-1]
        
        # Evaluate predictions
        evaluation = self.predictor.evaluatePredictions(predictions, actual)
        
        # Should have evaluation for all metrics
        self.assertIn('temperature', evaluation)
        self.assertIn('humidity', evaluation)
        self.assertIn('pressure', evaluation)
        
        # Each evaluation should have error metrics
        for metric_eval in evaluation.values():
            self.assertIn('absolute_error', metric_eval)
            self.assertIn('percent_error', metric_eval)


if __name__ == '__main__':
    unittest.main()
```

### Run Tests
- Testing using the `unittest` framework
  - In VS Code:
    - In the menu navigation on the left side, click on the beaker (mouseover the icons - hovering over the beaker will popup the name `Testing`)
    - All loaded tests will appear - click the `Refresh Testing` button to ensure they're all loaded.
      - NOTE: Tests must be in the top level `tests` path (within IPP_HOME) and each `unittest` module must begin with `test_` (e.g., `test_MyModule.py`)
    - Using the IDE's controls, execute the desired test(s). For this lab module, run the following:
      - `testPrepareTrainingData`
      - `testTrainModel`
      - `testPredictAfterTraining`
      - `testPredictWithoutTraining`
      - `testEvaluatePredictions`

- Testing the module directly
  - From within your IDE
    - Right click on your newly created module `test_SimpleWeatherPredictor.py` and click `run tests`
    - You should see output similar to that discussed in class

  - From the command line
    - Open a terminal and cd to your `IPP_HOME` path
    - Start your virtual environment (if not already running)
    - Be sure your PYTHONPATH is set correctly
    - Run the module
      - ```python -m unittest ./tests/labmodule09/test_SimpleWeatherPredictor.py```
    - You should see output similar to that discussed in class

### Sample output (yours may differ slightly)

```
Received test ids from temp file.
testEvaluatePredictions (labmodule09.test_SimpleWeatherPredictor.SimpleWeatherPredictorTest)
Test that prediction evaluation produces error metrics. ... KNN trained with 3 samples
Temperature predictor trained with 3 samples
KNN trained with 3 samples
Humidity predictor trained with 3 samples
KNN trained with 3 samples
Pressure predictor trained with 3 samples
Weather predictor training complete (iteration 1)
ok
testPredictAfterTraining (labmodule09.test_SimpleWeatherPredictor.SimpleWeatherPredictorTest)
Test that predictions can be made after training. ... ok
testPredictWithoutTraining (labmodule09.test_SimpleWeatherPredictor.SimpleWeatherPredictorTest)
Test that prediction fails gracefully when model is not trained. ... ok
testPrepareTrainingData (labmodule09.test_SimpleWeatherPredictor.SimpleWeatherPredictorTest)
Test that training data is prepared correctly from weather history. ... ok
testTrainModel (labmodule09.test_SimpleWeatherPredictor.SimpleWeatherPredictorTest)
Test that the predictor can be trained successfully. ... ok

----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
KNN trained with 3 samples
Temperature predictor trained with 3 samples
KNN trained with 3 samples
Humidity predictor trained with 3 samples
KNN trained with 3 samples
Pressure predictor trained with 3 samples
Weather predictor training complete (iteration 1)
Predictors not trained yet
KNN trained with 3 samples
Temperature predictor trained with 3 samples
KNN trained with 3 samples
Humidity predictor trained with 3 samples
KNN trained with 3 samples
Pressure predictor trained with 3 samples
Weather predictor training complete (iteration 1)
Finished running tests!
```
